Plugins/Public/base_plugin/CreateSolar.cpp

```diff
@@ -0,0 +1,290 @@
+#include "Main.h"
+
+
+void CreateSolar::CreateSolarCallout(SPAWN_SOLAR_STRUCT* info)
+{
+	uint spaceObjId;
+
+	pub::SpaceObj::SolarInfo si;
+	memset(&si, 0, sizeof(si));
+	si.iFlag = 4;
+	si.iArchID = info->solarArchetypeId;
+	si.iLoadoutID = info->loadoutArchetypeId;
+
+	si.iHitPointsLeft = -1;
+	si.iSystemID = info->iSystemId;
+	si.mOrientation = info->ori;
+	si.vPos = info->pos;
+	si.Costume.head = CreateID("pi_pirate2_head");
+	si.Costume.body = CreateID("pi_pirate8_body");
+	si.Costume.lefthand = 0;
+	si.Costume.righthand = 0;
+	si.Costume.accessories = 0;
+	si.iVoiceID = CreateID("atc_leg_m01");
+	strncpy_s(si.cNickName, sizeof(si.cNickName), info->nickname.c_str(), info->nickname.size());
+
+	struct PlayerData* pd = nullptr;
+	while (pd = Players.traverse_active(pd))
+	{
+		if(pd->iSystemID == info->iSystemId)
+			HkChangeIDSString(pd->iOnlineID, info->solar_ids, info->initialName);
+	}
+
+	// Set the base name
+	FmtStr infoname(info->solar_ids, 0);
+	infoname.begin_mad_lib(info->solar_ids); // scanner name
+	infoname.end_mad_lib();
+
+	FmtStr infocard(info->solar_ids, 0);
+	infocard.begin_mad_lib(info->solar_ids); // infocard
+	infocard.end_mad_lib();
+
+	pub::Reputation::Alloc(si.iRep, infoname, infocard);
+
+	SpawnSolar(spaceObjId, si);
+
+	pub::AI::SetPersonalityParams pers = MakePersonality();
+	pub::AI::SubmitState(spaceObjId, &pers);
+
+	info->iSpaceObjId = spaceObjId;
+
+	pub::SpaceObj::SetRelativeHealth(spaceObjId, info->percentageHp);
+	customSolarList.insert(spaceObjId);
+}
+
+void CreateSolar::SpawnSolar(unsigned int& spaceID, pub::SpaceObj::SolarInfo const& solarInfo)
+{
+	// hack server.dll so it does not call create solar packet send
+	char* serverHackAddress = (char*)hModServer + 0x2A62A;
+	char serverHack[] = { '\xEB' };
+	WriteProcMem(serverHackAddress, &serverHack, 1);
+
+	pub::SpaceObj::CreateSolar(spaceID, solarInfo);
+
+	uint dunno;
+	IObjInspectImpl* inspect;
+	if (GetShipInspect(spaceID, inspect, dunno))
+	{
+		CSolar* solar = (CSolar*)inspect->cobject();
+
+		// for every player in the same system, send solar creation packet
+		struct SOLAR_STRUCT
+		{
+			byte dunno[0x100];
+		};
+
+		SOLAR_STRUCT packetSolar;
+
+		char* address1 = (char*)hModServer + 0x163F0;
+		char* address2 = (char*)hModServer + 0x27950;
+
+		// fill struct
+		__asm
+		{
+			pushad
+			lea ecx, packetSolar
+			mov eax, address1
+			call eax
+			push solar
+			lea ecx, packetSolar
+			push ecx
+			mov eax, address2
+			call eax
+			add esp, 8
+			popad
+		}
+
+		struct PlayerData* pPD = 0;
+		while (pPD = Players.traverse_active(pPD))
+		{
+			if (pPD->iSystemID == solarInfo.iSystemID)
+				GetClientInterface()->Send_FLPACKET_SERVER_CREATESOLAR(pPD->iOnlineID, (FLPACKET_CREATESOLAR&)packetSolar);
+		}
+	}
+
+	// undo the server.dll hack
+	char serverUnHack[] = { '\x74' };
+	WriteProcMem(serverHackAddress, &serverUnHack, 1);
+}
+
+pub::AI::SetPersonalityParams CreateSolar::MakePersonality()
```

Comment:
With a tiny amount of code changes, you could make the personality block configurable and make it so certain platforms don't target certain classes and such. Currently the same personality is slapped on all solars.

Reply:
Merged a change supporting just that.